Detect user setup parallel backend

sritchie73 opened the pull request:
If the user has already registered a parallel backend for foreach in their R session (e.g. via `doMC::registerDoMC()` or `doParallel:registerDoParallel()`) then this is now detected, and that parallel backend used, instead of setting up a new one and clobbering the user's existing parallel backend.

Fix for https://github.com/privefl/bigsnpr/issues/251

sritchie73 committed:
Detect user setup parallel backend

If the user has already registered a parallel backend for foreach in their R session (e.g. via `doMC::registerDoMC()` or `doParallel:registerDoParallel()`) then this is now detected, and that parallel backend used, instead of setting up a new one and clobbering the user's existing parallel backend.

Fix for https://github.com/privefl/bigsnpr/issues/251

privefl commented:
Thanks for this.
I was thinking of something like this: https://github.com/privefl/bigstatsr/issues/118.

sritchie73 commented:
I can't really comment on the `doFuture` package as I haven't played around with it.

I think an overarching issue is that theres currently a bit of "gotcha" for users of your family of packages: functions in bigsnpr (e.g the LDpred2 functions) and bigstatsr (e.g big_prodMat) all have an `ncores` argument, but use different methods of paralellism, and currently have different side effects on the users R session.

E.g. big_prodMat (and other functions in bigstatsr which call register_parallel) will overwrite any existing user backend, potentially causing crashes downstream for the use if they try to use foreach %dopar% after running these functions. Conversely, the main ldpred2 functions in bigsnpr don't have this effect, because their parallel code is in C++. The change I propose fixes this by detecting whether a parallel backend for foreach is already registered, and using this instead.

If you're proposing instead to have `ncores = NULL` to mean "use an existing parallel backend instead of registering a new one", then without diving into the source code, I would also expect to pass `ncores = NULL` to the ldpred2 functions (e.g. `snp_ldpred2_grid()`).

privefl commented:
Yes, there are many different parallel strategies used in the packages, because different approaches work best in different situations. I should probably document what function uses what strategy.

For using an already registered parallel backend, I would like users to be explicit about it (therefore using something like `NULL`) as they could forget about it and instead think they are using the default `ncores = 1`. But that would work only for the functions using foreach.

sritchie73 commented:
In that case, I would suggest outputing a warning message if a pre-registered backend is detected, but the user hasn't set ncores = NULL? 

E.g. something like:

```
if (foreach::getDoParWorkers() > 1 && !is.null(ncores)) {
  warning("User registered parallel backend detected: 'ncores' should be set to NULL
}
```

privefl commented:
I'll think about it. Thanks.

privefl commented:
I don't think your solution would work because of:
```r
cl <- makeCluster(4)
registerDoParallel(cl)
stopCluster(cl)
foreach::getDoParWorkers()   # 4
```

sritchie73 commented:
That seems like a bug with doParallel to be honest - I would have thought stopCluster() would have cleaned up properly.

From this [SO post](https://stackoverflow.com/questions/25097729/un-register-a-doparallel-cluster) it looks like you should also an additional `foreach::registerDoSeq()` in your on.exit() call after stopCluster().

sritchie73 committed:
Deregister doParallel cluster with foreach

After stopCluster() also need to let foreach know parallel backend has been removed.

privefl commented:
I think I will switch to using {future}, where you can properly save a previous `plan()` and apply it again when exiting the function.

It should take a few days, I will keep you updated in the original issue.
